### PR TITLE
v3.0.0

### PR DESCRIPTION
**Pull request by jakekrell**
- Initialize model with only user-defined hyperparameters, with the rest set to default values
- Automatic formatting of inputs and data, including normalization of inputs
- Keyword 'train' for automatic test/train splits
- Keywords added to model.coverage3() for user control of plotting
- Storage of various info such as inputs, data, normalization scale factors, etc. as attributes of model
- Function model.clear() for removing attributes from model that are not hyperparameters, to enable sweeping / iterating through datasets without needing to create a new FoKL class

### CODE REVIEW

File: README.md
Comment (derek-slack) [inline README.md:1]: Line 66 - Review difference between model.inputs and test and train data, model inputs contains entire data set. Should only run coverage on the test and train inputs. Just double check this functionality on the test scripts as well. Let me know if you want clarification

File: README.md [hunk 1]
```diff
@@ -3,55 +3,124 @@ Karhunen Loève decomposed Gaussian processes with forward variable
...
+As a side note, the following attributes were added to your FoKL class 'model' after calling 'fit' which may be useful during user post-processing:
+```
+model.inputs         == all normalized inputs w/o outliers (i.e., model.traininputs plus model.testinputs)
+model.data           == all data w/o outliers (i.e., model.traindata plus model.testdata)
+
+model.betas          == betas
+model.mtx            == mtx
+model.evs            == evs
+
+model.rawinputs      == all normalized inputs w/ outliers == user's 'inputs' but normalized and formatted
```
Comment (derek-slack) [inline README.md:79]: Line 79 - if you mention outliers I think you should mention how you define outliers and the outlier methods in here

File: README.md [hunk 2]
```diff
@@ -3,55 +3,124 @@ Karhunen Loève decomposed Gaussian processes with forward variable
...
-and then used as 
 
+## Integration
+FoKL can be used to model state derivatives and thus contains an integration method of these states using an RK4. Due to each state being modeled independently, the same functionality cannot be used. For the case of two states with the same inputs:
 ```
-T,Y = FoKLRoutines.FoKL.GP_Integrate([np.mean(betas1,axis=0),np.mean(betas2,axis=0)], [mtx1,mtx2], utest, norms, phis, start, stop, ic, stepsize, used_inputs)
+data = [state1, state2]
+betas = []
+mtx = []
+for i in range(2):
```
Comment (derek-slack) [inline README.md:106]: Line 106 - IMO simpler is better sometimes for readability, is it necessary to do a for loop for just a two state model? Just might be easier to show fitting two separate models and putting them both in.

Comment (derek-slack) [reply README.md:106]: You could then cal in gp integrate (model1.betas) as opposed to having to redefine it as a global variable

File: src/FoKL/FoKLRoutines.py [hunk 3]
```diff
@@ -356,6 +813,9 @@ def gibbs(inputs, data, phis, Xin, discmtx, a, b, atau, btau, draws):
                 Lamb_tausqd_inv = np.diag(1 / np.diag(Lamb_tausqd))
 
                 mun = Q.dot(Lamb_tausqd_inv).dot(np.transpose(Q)).dot(Xty)
+                if mun.ndim == 1: # if mun.shape == (number,) != (number,1), then add new axis
+                    mun = mun[:, np.newaxis]
+                    warnings.warn("'mun' was made into (n,1) column vector from single list (n,). It is unclear why this was not already the case.",category=UserWarning)
```
Comment (derek-slack) [inline src/FoKL/FoKLRoutines.py:818]: Line 818 - How often does this happen? Is it unclear why this is the case?

File: src/FoKL/FoKLRoutines.py [hunk 4]
```diff
@@ -263,6 +719,7 @@ def gibbs(inputs, data, phis, Xin, discmtx, a, b, atau, btau, draws):
             # outputs for each set of inputs
             minp, ninp = np.shape(inputs)
             phi_vec = []
+            # CHECK BELOW !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
Comment (derek-slack) [inline src/FoKL/FoKLRoutines.py:722]: Line 722 + 738 - Remove Irrelevant Comments

File: Examples/Sigmoid/SigmoidTest.py [hunk 5]
```diff
@@ -1,150 +1,65 @@
...
-        'data' - test set results
-        'draws' - number of beta models used from model.fit()
-        plotting binary - 1 = show plot, 0 = do not show plot
-        
-    outputs:
-        'meen' - indexed predicted value
-        'bounds' - 95% confidence interval for each point
-        'rmse' - root mean square error
-"""
+from src.FoKL import FoKLRoutines # from FoKL import FoKLRoutines
```
Comment (derek-slack) [inline Examples/Sigmoid/SigmoidTest.py:2]: I don't believe from src.FoKL is necessary here FoKL package imports all the modules, should just be able to do 

```
from FoKL import FoKLRoutines
```

**Review by derek-slack — CHANGES_REQUESTED** [review]
Check Comments that I made. Functionality question:

- Can I create model object with no kwargs then define them as model.a = 1 and so on?

- should necessary inputs be kwargs? What do other ML packages do?

File: Examples/Sigmoid/SigmoidTest.py (hunk 5 above)
Comment (derek-slack) [reply Examples/Sigmoid/SigmoidTest.py:2]: double check this before committing

File: README.md
Comment (jakekrell) [reply README.md:1]: - correct, model.inputs and model.data contains entire dataset that user entered into fit --- by default (as of now), coverage evaluates [train+test] inputs
- documentation was wrong -- double checked code and it is ok -- updated now
- (note data is required in coverage for rmse and plotting ... in future might be good to add functionality for only calculating meens and bounds without returning rmse or plot)

File: README.md (hunk 2 above)
Comment (jakekrell) [reply README.md:106]: i could see pros/cons for both approaches --- imo it'd be easier for a user to take the for loop (even if it's only 2 states here) and generalize it to many states or to a sweep of hypers, but not the other way around

i figure model1 and model2 would be obvious without an example (or at least all the tools are available from "Setup" section), but a loop might not be as obvious so this is kinda acting as a two-birds one-stone for showing loops

BUT, maybe an extra section after "Integration" called "Sweeps" or something like that? That way we keep readability (by using model1, model2) but also have docs for how to loop

(also ... forgot to add "model.clear()" at the end of the loop which is the main piece of syntax to convey to user about using loops)

File: src/FoKL/FoKLRoutines.py (hunk 3 above)
Comment (jakekrell) [reply src/FoKL/FoKLRoutines.py:818]: does not happen with Sigmoid, GP Int, or F16

it could be deleted but might be good to keep in case of future de-bugging --- was likely originating from an old problem in auto_cleanData() which is now resolved

-------------------

what does happen is:
      (which isn't a problem but just a note for developing, so could be deleted from Routines or just import warnings and turn off UserWarnings)

Sigmoid ---> 'inputs' was transposed, ignore is more datapoints than input variables

GP Int -----> 'data' was made into (n,1) column vector ...

F16 ---> 'inputs' transposed, ignore .. ; 'data' transposed, ignore ...

Comment (jakekrell) [reply src/FoKL/FoKLRoutines.py:818]: personally i like the "inputs/data was transposed, ignore if datapoints > variables" cause just in case someone tries to fit 10 inputs variables x 9 datapoints/timepoints, it'll let them know that their data was flipped

the "(n,) to (n,1)" is kinda pointless practically but i'm not 100% why it happens so might be worth keeping

File: Examples/Sigmoid/SigmoidTest.py (hunk 5 above)
Comment (jakekrell) [reply Examples/Sigmoid/SigmoidTest.py:2]: correct -- my mistake leaving that in. fixed

**Comment by jakekrell** [comment]
- Q: Can I create model object with no kwargs then define them as model.a = 1 and so on?
- A: yes. --- by default model.a=4 if you do not use kwargs, then model.a=1 will update that just fine

- Q: should necessary inputs be kwargs? What do other ML packages do?
- A: what would necessary inputs be? btau and b get defined in fit (if not entered as kwarg) based on sigmasq of data (so not needed from user), but other than that i can't think what couldn't be set as a default value ........... to be fair the default values might be awful for the given userinputs and userdata, but i figure at least a relatively default value gets them started without needed to input anything besides inputs/data .......................................... not sure what's standard for other packages. have any suggestions on ones to check out?

thanks for the helpful feedback/editing! lmk if you see anything else
